### PR TITLE
spline #221 - add unit test for UpgraderV4

### PR DESCRIPTION
**Pull request by wajda**

### CODE REVIEW

File: persistence/mongo/src/test/scala/za/co/absa/spline/persistence/mongo/dao/MutableLineageUpgraderV4Spec.scala [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+import scala.collection.JavaConverters._
+import scala.concurrent.{ExecutionContext, Future}
+
+class MutableLineageUpgraderV4Spec extends AsyncFlatSpec with Matchers {
+
+  behavior of "MutableLineageUpgraderV4"
+
+  it should "add empty metrics to Write operation when ones are missing" in {
+    val testUpgrader = (new VersionedLineageDAODummy with MutableLineageUpgraderV4).upgrader.get
```
[the remark this answers is not in the document]
Comment (wajda) [reply persistence/mongo/src/test/scala/za/co/absa/spline/persistence/mongo/dao/MutableLineageUpgraderV4Spec.scala]: it's OK in this case as it's a part of construction of an instance that is a subject for the test.

File: persistence/mongo/src/test/scala/za/co/absa/spline/persistence/mongo/dao/MutableLineageUpgraderV4Spec.scala [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  }
+
+  trait VersionedLineageDAODummy extends VersionedLineageDAO {
+    override def save(lineage: DBObject)(implicit e: ExecutionContext): Future[Unit] = noImpl
+
+    override def loadByDatasetId(dsId: UUID, overviewOnly: Boolean)(implicit ec: ExecutionContext): Future[Option[DBObject]] = noImpl
+
+    override def searchDataset(path: String, applicationId: String)(implicit ec: ExecutionContext): Future[Option[UUID]] = noImpl
+
+    override def getLastOverwriteTimestampIfExists(path: String)(implicit ec: ExecutionContext): Future[Option[Timestamp]] = noImpl
```
[the remark this answers is not in the document]
Comment (lukasnalezenec) [reply persistence/mongo/src/test/scala/za/co/absa/spline/persistence/mongo/dao/MutableLineageUpgraderV4Spec.scala]: We could turn off this codacy bot rule

File: persistence/mongo/src/test/scala/za/co/absa/spline/persistence/mongo/dao/MutableLineageUpgraderV4Spec.scala [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+    override def findDatasetIdsByPathSince(path: String, since: Timestamp)(implicit ec: ExecutionContext): Future[CloseableIterable[UUID]] = noImpl
+
+    override def findByInputId(datasetId: UUID, overviewOnly: Boolean)(implicit ec: ExecutionContext): Future[CloseableIterable[DBObject]] = noImpl
+
+    override def findDatasetDescriptors(maybeText: Option[String], pageRequest: DataLineageReader.PageRequest)(implicit ec: ExecutionContext): Future[CloseableIterable[DBObject]] = noImpl
```
[the remark this answers is not in the document]
Comment (lukasnalezenec) [reply persistence/mongo/src/test/scala/za/co/absa/spline/persistence/mongo/dao/MutableLineageUpgraderV4Spec.scala]: IMHO we should split this line.

File: persistence/mongo/src/test/scala/za/co/absa/spline/persistence/mongo/dao/MutableLineageUpgraderV4Spec.scala [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        "_id" -> "ln_123",
+        "attributes" -> new BasicDBList,
+        "operations" -> new BasicDBList {
+          add(new BasicDBObject(Map("_typeHint" -> "za.co.absa.spline.model.op.Write").asJava))
+          add(new BasicDBObject(Map("_typeHint" -> "za.co.absa.spline.model.op.Other").asJava))
+        }
+      ).asJava)
+
+    for (resultedLineageDBOv4 <- testUpgrader(fakeLineageDBOv3)) yield {
+      val operationsDBOs = (resultedLineageDBOv4 get "operations").asInstanceOf[BasicDBList]
```
Comment (lukasnalezenec) [inline persistence/mongo/src/test/scala/za/co/absa/spline/persistence/mongo/dao/MutableLineageUpgraderV4Spec.scala]: Nips: I would prefer to call  resultedLineageDBOv4.get("operations")

**Pull request merged**